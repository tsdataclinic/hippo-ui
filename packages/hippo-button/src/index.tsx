import { ThemeContext } from '@hippo/theme-provider';
import { useContext, useEffect } from 'react';
import * as React from 'react';

type HTMLButtonProps = React.ComponentPropsWithoutRef<'button'>;
type HTMLButtonRef = React.ElementRef<'button'>;

export interface ButtonProps extends HTMLButtonProps {
  children: React.ReactNode;
}
const NAME = 'HippoButton';
export const Button = React.forwardRef<HTMLButtonRef, ButtonProps>(
  (props: ButtonProps, forwardedRef: React.ForwardedRef<HTMLButtonRef>) => {
    const { children, type, ...passThroughProps } = props;

    const { highlightedComponents, registerComponentName, theme } =
      useContext(ThemeContext);

    useEffect(() => {
      console.log('Registering ' + NAME);
      registerComponentName(NAME);
    }, []);
    const configs = theme.componentSpecificConfigs[NAME];
    const isHighlighted = new Set(highlightedComponents).has(NAME);

    return (
      <button
        ref={forwardedRef}
        type={type ?? 'button'}
        {...passThroughProps}
        style={{
          color: theme.color,
          fontSize: theme.fontSize,
          ...configs,
          backgroundColor: isHighlighted ? 'Yellow' : 'inherit',
        }}
      >
        {children}
      </button>
    );
  },
);

Button.displayName = 'Button';
