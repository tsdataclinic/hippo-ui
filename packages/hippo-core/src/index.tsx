export { Button, type ButtonProps } from "./Button";
export { InputText, type InputTextProps } from "./InputText";
