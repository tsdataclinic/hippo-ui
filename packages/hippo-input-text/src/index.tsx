import * as React from 'react';
import { useTextField, type AriaTextFieldOptions } from '@react-aria/textfield';
import { ThemeContext } from '@hippo/theme-provider';
import { useComposedRefs } from '@hippo/utils';

type HTMLInputRef = React.ElementRef<'input'>;

export interface InputTextProps extends AriaTextFieldOptions<'input'> {
  /**
   * The value to set for the InputText. If no value is provided, then the
   * InputText is treated as uncontrolled.
   */
  value?: string;
}

const NAME = 'HippoInputText';

/**
 * An InputText component.
 * Is uncontrolled by default until a `value` is passed.
 */
export const InputText = React.forwardRef<HTMLInputRef, InputTextProps>(
  (props: InputTextProps, forwardedRef: React.ForwardedRef<HTMLInputRef>) => {
    const inputRef = React.useRef<HTMLInputElement | null>(null);
    const composedRef = useComposedRefs(forwardedRef, inputRef);
    const { inputProps } = useTextField(props, composedRef);
    const { value, ...passThroughInputProps } = inputProps;

    const { highlightedComponents, registerComponentName, theme } =
      React.useContext(ThemeContext);

    React.useEffect(() => {
      console.log('Registering ' + NAME);
      registerComponentName(NAME);
    }, []);
    const configs = theme.componentSpecificConfigs[NAME];
    const isHighlighted = new Set(highlightedComponents).has(NAME);

    return (
      <input
        value={value}
        {...passThroughInputProps}
        ref={forwardedRef}
        type="text"
        style={{
          color: theme.color,
          fontSize: theme.fontSize,
          ...configs,
          backgroundColor: isHighlighted ? 'Yellow' : 'inherit',
        }}
      />
    );
  },
);

InputText.displayName = 'InputText';
