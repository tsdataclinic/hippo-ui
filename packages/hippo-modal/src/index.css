.hippo-modal__dialog-content {
  background-color: red;
}
