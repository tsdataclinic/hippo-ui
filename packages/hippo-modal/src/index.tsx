import * as React from 'react';
import * as Dialog from '@radix-ui/react-dialog';
import styled, { keyframes } from 'styled-components';

export interface ModalProps {
  children: React.ReactNode;
  isOpen: boolean;
  onDismiss: () => void;
  title: string;
}

const overlayShow = keyframes({
  '0%': { opacity: 0 },
  '100%': { opacity: 1 },
});

const ModalOverlay = styled(Dialog.Overlay)`
  background-color: rgba(0, 0, 0, 0.05);
  position: fixed;
  inset: 0;
  '@media (prefers-reduced-motion: no-preference)': {
    animation: ${overlayShow} 150ms cubic-bezier(0.16, 1, 0.3, 1),
  },
`;

const ModalContent = styled(Dialog.Content)`
  background: white;
  box-shadow: hsl(206 22% 7% / 35%) 0px 10px 38px -10px,
    hsl(206 22% 7% / 20%) 0px 10px 20px -15px;
  left: 50%;
  maxheight: 85vh;
  maxwidth: 450px;
  padding: 25;
  position: fixed;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 90vw;

  &:focus {
    outline: none;
  }
`;

export function Modal({
  children,
  isOpen,
  onDismiss,
  title,
}: ModalProps): JSX.Element {
  const onOpenChange = React.useCallback(
    (open: boolean) => {
      if (!open) {
        onDismiss();
      }
    },
    [onDismiss],
  );

  return (
    <Dialog.Root open={isOpen} onOpenChange={onOpenChange}>
      <Dialog.Portal>
        <ModalOverlay />
        <ModalContent className="testing">
          <Dialog.Title>{title}</Dialog.Title>
          {children}
          <Dialog.Close asChild>
            <button>Close</button>
          </Dialog.Close>
        </ModalContent>
      </Dialog.Portal>
    </Dialog.Root>
  );
}

Modal.displayName = 'Modal';
