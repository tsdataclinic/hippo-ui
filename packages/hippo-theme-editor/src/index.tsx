import { ThemeContext } from '@hippo/theme-provider';
import { Modal } from '@hippo/modal';
import { GlobalHotKeys } from 'react-hotkeys';
import * as React from 'react';
import StartButton from './StartButton';

export function ThemeEditor(): JSX.Element {
  const [isThemeEditorOpen, setIsThemeEditorOpen] =
    React.useState<boolean>(false);

  const {
    setColor,
    setFontSize,
    setPadding,
    theme,
    setComponentSpecificConfigs,
    setHighlightedComponents,
  } = React.useContext(ThemeContext);
  const [selectedComponentName, setSelectedComponentName] = React.useState<
    string | undefined
  >(undefined);
  const selectedComponentConfig = React.useMemo(
    () =>
      (selectedComponentName &&
        theme.componentSpecificConfigs[selectedComponentName]) ??
      null,
    [theme, selectedComponentName],
  );
  const [isDebugging, setIsDebugging] = React.useState<boolean>(false);
  React.useEffect(() => {
    if (selectedComponentName) {
      setHighlightedComponents([selectedComponentName]);
    } else {
      setHighlightedComponents([]);
    }
  }, [selectedComponentName]);

  return (
    <>
      <GlobalHotKeys
        keyMap={{
          TOGGLE: 'e',
        }}
        handlers={{
          TOGGLE: keyEvent => {
            if (keyEvent) {
              keyEvent.preventDefault();
            }
            setIsThemeEditorOpen((v: boolean) => !v);
          },
        }}
      />
      <Modal
        isOpen={isThemeEditorOpen}
        onDismiss={() => {
          setIsThemeEditorOpen(false);
          setHighlightedComponents([]);
        }}
        title="Hippo Style Editor"
      >
        <div
          style={{
            border: '1px solid teal',
            padding: 40,
            position: 'relative',
          }}
        >
          <div style={{ width: '100%', textAlign: 'left' }}>
            Registered components:{' '}
            {[...Object.keys(theme.componentSpecificConfigs)].map(
              (name, index) => {
                const config = theme.componentSpecificConfigs[name];
                return (
                  <span key={name}>
                    {index !== 0 && <span style={{ marginRight: 2 }}>,</span>}
                    <span
                      style={{ color: 'teal' }}
                      onMouseEnter={() => {
                        if (!selectedComponentName) {
                          setHighlightedComponents([name]);
                        }
                      }}
                      onMouseLeave={() => {
                        if (!selectedComponentName) {
                          setHighlightedComponents([]);
                        }
                      }}
                    >
                      {name}
                    </span>
                  </span>
                );
              },
            )}
          </div>
          <hr />
          <div style={{ width: '100%', textAlign: 'left', marginBottom: 8 }}>
            <span>Apply style changes to </span>
            <select
              value={selectedComponentName}
              onChange={ev => {
                if (ev.target.value === 'all') {
                  setSelectedComponentName(undefined);
                } else {
                  setSelectedComponentName(ev.target.value);
                }
              }}
            >
              <option value="all">Global</option>
              {[...Object.keys(theme.componentSpecificConfigs)].map(
                componentName => {
                  return (
                    <option key={componentName} value={componentName}>
                      {componentName}
                    </option>
                  );
                },
              )}
            </select>
          </div>
          <div style={{ display: 'flex' }}>
            <span>Text Color</span>
            <span style={{ marginLeft: 8 }}>
              <input
                type="color"
                value={
                  selectedComponentName
                    ? selectedComponentConfig.color
                    : theme.color
                }
                onChange={e => {
                  if (selectedComponentName) {
                    setComponentSpecificConfigs(selectedComponentName, {
                      color: e.target.value,
                    });
                  } else {
                    setColor(e.target.value);
                  }
                }}
              />
            </span>
          </div>
          <div style={{ display: 'flex' }}>
            <span>Font Size</span>
            <span style={{ marginLeft: 8 }}>
              <input
                type="range"
                min="9"
                max="96"
                value={
                  selectedComponentName
                    ? selectedComponentConfig.fontSize
                    : theme.fontSize
                }
                className="slider"
                onChange={e => {
                  if (selectedComponentName) {
                    setComponentSpecificConfigs(selectedComponentName, {
                      fontSize: parseInt(e.target.value, 10),
                    });
                  } else {
                    setFontSize(parseInt(e.target.value, 10));
                  }
                }}
              />
            </span>
            <span style={{ marginLeft: 4 }}>
              {selectedComponentName
                ? selectedComponentConfig.fontSize
                : theme.fontSize}
            </span>
          </div>

          <div style={{ display: 'flex' }}>
            <span>Padding - Small</span>
            <span style={{ marginLeft: 8 }}>
              <input
                type="range"
                min="1"
                max="100"
                value={
                  typeof theme.paddings.sm === 'string'
                    ? parseInt(theme.paddings.sm, 10)
                    : theme.paddings.sm
                }
                className="slider"
                onChange={e => {
                  if (selectedComponentName) {
                    setComponentSpecificConfigs(selectedComponentName, {
                      paddings: {
                        ...theme.componentSpecificConfigs[selectedComponentName]
                          .paddings,
                        sm: parseInt(e.target.value, 10),
                      },
                    });
                  } else {
                    setPadding('sm', parseInt(e.target.value, 10));
                  }
                }}
              />
            </span>
            <span style={{ marginLeft: 4 }}>{theme.paddings.sm}</span>
          </div>

          <div style={{ display: 'flex' }}>
            <span>Padding - Medium</span>
            <span style={{ marginLeft: 8 }}>
              <input
                type="range"
                min="1"
                max="100"
                value={
                  typeof theme.paddings.md === 'string'
                    ? parseInt(theme.paddings.md, 10)
                    : theme.paddings.md
                }
                className="slider"
                onChange={e => {
                  if (selectedComponentName) {
                    setComponentSpecificConfigs(selectedComponentName, {
                      paddings: {
                        ...theme.componentSpecificConfigs[selectedComponentName]
                          .paddings,
                        md: parseInt(e.target.value, 10),
                      },
                    });
                  } else {
                    setPadding('md', parseInt(e.target.value, 10));
                  }
                }}
              />
            </span>
            <span style={{ marginLeft: 4 }}>{theme.paddings.md}</span>
          </div>

          <div style={{ display: 'flex' }}>
            <span>Padding - Large</span>
            <span style={{ marginLeft: 8 }}>
              <input
                type="range"
                min="1"
                max="100"
                value={
                  typeof theme.paddings.lg === 'string'
                    ? parseInt(theme.paddings.lg, 10)
                    : theme.paddings.lg
                }
                className="slider"
                onChange={e => {
                  if (selectedComponentName) {
                    setComponentSpecificConfigs(selectedComponentName, {
                      paddings: {
                        ...theme.componentSpecificConfigs[selectedComponentName]
                          .paddings,
                        lg: parseInt(e.target.value, 10),
                      },
                    });
                  } else {
                    setPadding('lg', parseInt(e.target.value, 10));
                  }
                }}
              />
            </span>
            <span style={{ marginLeft: 4 }}>{theme.paddings.lg}</span>
          </div>

          {isDebugging && (
            <>
              <hr />
              <pre style={{ width: '100%', textAlign: 'left' }}>
                {JSON.stringify(theme, null, 2)}
              </pre>
            </>
          )}
          <button
            style={{ position: 'absolute', bottom: 0, right: 0 }}
            onClick={() => {
              setIsDebugging(v => !v);
            }}
          >
            Debug
          </button>
        </div>
      </Modal>
      <StartButton
        onClick={() => {
          setIsThemeEditorOpen(true);
        }}
      />
    </>
  );
}
