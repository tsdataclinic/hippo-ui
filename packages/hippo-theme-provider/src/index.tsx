import * as React from 'react';
export { ThemeContext } from './ThemeContext';
export { ThemeProvider } from './ThemeProvider';
export { ThemeEditor } from './ThemeEditor';
export { default as MyButton } from './MyButton';
export { default as MyDiv } from './MyDiv';
