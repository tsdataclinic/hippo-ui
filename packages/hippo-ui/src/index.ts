export {Button} from '@hippo/button';
export {InputText} from '@hippo/input-text';
export {ThemeContext, ThemeProvider, ThemeEditor} from '@hippo/theme-provider';