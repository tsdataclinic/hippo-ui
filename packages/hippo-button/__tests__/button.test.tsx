import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Button } from '../src/index';

describe('hippo-button', () => {
  test('should render a single child as expected', () => {
    render(<Button>Inner Text</Button>);

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  test('should render multiple children as expected', () => {
    render(
      <Button>
        <span>Child 1</span>
        <span>Child 2</span>
      </Button>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });
});
