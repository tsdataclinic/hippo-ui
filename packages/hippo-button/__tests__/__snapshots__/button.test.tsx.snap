// Vitest Snapshot v1

exports[`hippo-button > should render a single child as expected 1`] = `
<button>
  Inner Text
</button>
`;

exports[`hippo-button > should render multiple children as expected 1`] = `
<button>
  <span>
    Child 1
  </span>
  <span>
    Child 2
  </span>
</button>
`;
